src/timezones/types.jl

```diff
@@ -0,0 +1,212 @@
+
+# import Base.Dates: UTInstant, DateTime, TimeZone, Millisecond
+using Base.Dates
+import Base.Dates: value
+
+# Note: The Olsen Database rounds offset precision to the nearest second
+# See "America/New_York" notes for an example.
+
+# Using type Symbol instead of AbstractString for name since it
+# gets us ==, and hash for free.
+immutable FixedTimeZone <: TimeZone
+    name::Symbol
+    utc_offset::Second  # Standard offset from UTC
+    dst_offset::Second  # Addition offset applied to UTC offset
+
+    function FixedTimeZone(name::Symbol, utc_offset::Second, dst_offset::Second=Second(0))
+        new(name, utc_offset, dst_offset)
+    end
+end
+
+function FixedTimeZone(name::String, utc_offset::Int, dst_offset::Int=0)
+    FixedTimeZone(symbol(name), Second(utc_offset), Second(dst_offset))
+end
+
+function FixedTimeZone(s::String)
+    m = match(r"^([+-]?\d{2})\:?(\d{2})(?:\:(\d{2}+))?$", s)
+    m == nothing && error("Unrecognized timezone: $s")
+
+    # TODO: Somehow this is not working in the package but works in the REPL
+    values = map(n -> n == nothing ? 0 : Base.parse(Int, n), m.captures)
+
+    if values[3] == 0
+        name = @sprintf("UTC%+03d:%02d", values[1:2]...)
+    else
+        name = @sprintf("UTC%+03d:%02d:%02d", values...)
+    end
+
+    if values[1] < 0
+        for i in 2:length(values)
+            values[i] = -values[i]
+        end
+    end
+
+    offset = values[1] * 3600 + values[2] * 60 + values[3]
+    return FixedTimeZone(name, offset)
+end
+
+immutable Transition
+    utc_datetime::DateTime  # Instant where new zone applies
+    zone::FixedTimeZone
+end
+
+Base.isless(x::Transition,y::Transition) = isless(x.utc_datetime,y.utc_datetime)
+
+# function Base.string(t::Transition)
+#     return "$(t.utc_datetime), $(t.zone.name), $(t.zone.offset)"
+# end
+
+"""
+A TimeZone that has a variable offset from UTC.
+"""
+immutable VariableTimeZone <: TimeZone
+    name::Symbol
+    transitions::Vector{Transition}
+end
+
+function VariableTimeZone(name::String, transitions::Vector{Transition})
+    return VariableTimeZone(symbol(name), transitions)
+end
+
+immutable ZonedDateTime <: TimeType
+    utc_datetime::DateTime
+    timezone::TimeZone
```

Comment:
I'm conflicted on this. First off due to all `TimeZone` types containing a `Symbol` the objects will never be inlined since `isbits` will never be true. Second, the timezone field is only around for when we perform calculations with the `ZonedDateTime` and is only around as a reference so it kinda makes sense to have it as a reference.

Are there any other reasons to parameterize this Type? I'm mostly just curious if I'm not thinking of some feature in Julia.